profiler causing chunk nullpointerexception

Java version 17
Custom TacoSpigot 1.8.8 fork

```
[01:19:21 WARN]: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.NullPointerException
[01:19:21 WARN]:        at me.lucko.spark.common.platform.world.AsyncWorldInfoProvider.get(AsyncWorldInfoProvider.java:68)
[01:19:21 WARN]:        at me.lucko.spark.common.platform.world.AsyncWorldInfoProvider.getCounts(AsyncWorldInfoProvider.java:84)
[01:19:21 WARN]:        at me.lucko.spark.common.sampler.window.WindowStatisticsCollector.measure(WindowStatisticsCollector.java:159)
[01:19:21 WARN]:        at me.lucko.spark.common.sampler.window.WindowStatisticsCollector.lambda$measureNow$0(WindowStatisticsCollector.java:108)
[01:19:21 WARN]:        at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
[01:19:21 WARN]:        at me.lucko.spark.common.sampler.window.WindowStatisticsCollector.measureNow(WindowStatisticsCollector.java:108)
[01:19:21 WARN]:        at me.lucko.spark.common.sampler.async.AsyncSampler.rotateProfilerJob(AsyncSampler.java:111)
[01:19:21 WARN]:        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
[01:19:21 WARN]:        at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
[01:19:21 WARN]:        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
[01:19:21 WARN]:        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[01:19:21 WARN]:        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[01:19:21 WARN]:        at java.base/java.lang.Thread.run(Thread.java:833)
[01:19:21 WARN]: Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException
[01:19:21 WARN]:        at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
[01:19:21 WARN]:        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
[01:19:21 WARN]:        at me.lucko.spark.common.platform.world.AsyncWorldInfoProvider.get(AsyncWorldInfoProvider.java:66)
[01:19:21 WARN]:        ... 12 more
[01:19:21 WARN]: Caused by: java.lang.NullPointerException
```

and

```
[00:59:21 WARN]: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot invoke "org.bukkit.Chunk.getTileEntities()" because "chunk" is null
[00:59:21 WARN]:        at me.lucko.spark.common.platform.world.AsyncWorldInfoProvider.get(AsyncWorldInfoProvider.java:68)
[00:59:21 WARN]:        at me.lucko.spark.common.platform.world.AsyncWorldInfoProvider.getCounts(AsyncWorldInfoProvider.java:84)
[00:59:21 WARN]:        at me.lucko.spark.common.sampler.window.WindowStatisticsCollector.measure(WindowStatisticsCollector.java:159)
[00:59:21 WARN]:        at me.lucko.spark.common.sampler.window.WindowStatisticsCollector.lambda$measureNow$0(WindowStatisticsCollector.java:108)
[00:59:21 WARN]:        at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
[00:59:21 WARN]:        at me.lucko.spark.common.sampler.window.WindowStatisticsCollector.measureNow(WindowStatisticsCollector.java:108)
[00:59:21 WARN]:        at me.lucko.spark.common.sampler.async.AsyncSampler.rotateProfilerJob(AsyncSampler.java:111)
[00:59:21 WARN]:        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
[00:59:21 WARN]:        at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
[00:59:21 WARN]:        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
[00:59:21 WARN]:        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[00:59:21 WARN]:        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[00:59:21 WARN]:        at java.base/java.lang.Thread.run(Thread.java:833)
[00:59:21 WARN]: Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot invoke "org.bukkit.Chunk.getTileEntities()" because "chunk" is null
[00:59:21 WARN]:        at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
[00:59:21 WARN]:        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
[00:59:21 WARN]:        at me.lucko.spark.common.platform.world.AsyncWorldInfoProvider.get(AsyncWorldInfoProvider.java:66)
[00:59:21 WARN]:        ... 12 more
[00:59:21 WARN]: Caused by: java.lang.NullPointerException: Cannot invoke "org.bukkit.Chunk.getTileEntities()" because "chunk" is null
[00:59:21 WARN]:        at me.lucko.spark.bukkit.BukkitWorldInfoProvider.pollCounts(BukkitWorldInfoProvider.java:74)
[00:59:21 WARN]:        at me.lucko.spark.common.platform.world.AsyncWorldInfoProvider.lambda$async$0(AsyncWorldInfoProvider.java:54)
[00:59:21 WARN]:        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
[00:59:21 WARN]:        at org.bukkit.craftbukkit.v1_8_R3.scheduler.CraftTask.run(CraftTask.java:54)
[00:59:21 WARN]:        at org.bukkit.craftbukkit.v1_8_R3.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:332)
[00:59:21 WARN]:        at net.minecraft.server.v1_8_R3.MinecraftServer.B(MinecraftServer.java:876)
[00:59:21 WARN]:        at net.minecraft.server.v1_8_R3.DedicatedServer.B(DedicatedServer.java:387)
[00:59:21 WARN]:        at net.minecraft.server.v1_8_R3.MinecraftServer.A(MinecraftServer.java:786)
[00:59:21 WARN]:        at net.minecraft.server.v1_8_R3.MinecraftServer$2.run(MinecraftServer.java:674)
[00:59:21 WARN]:        ... 6 more
```

--- a/spark-bukkit/src/main/java/me/lucko/spark/bukkit/BukkitWorldInfoProvider.java
+++ b/spark-bukkit/src/main/java/me/lucko/spark/bukkit/BukkitWorldInfoProvider.java
@@ -27,6 +27,7 @@ import me.lucko.spark.common.platform.world.WorldInfoProvider;
 import org.bukkit.Chunk;
 import org.bukkit.Server;
 import org.bukkit.World;
+import org.bukkit.block.BlockState;
 import org.bukkit.entity.Entity;
 import org.bukkit.entity.EntityType;
 
@@ -71,7 +72,8 @@ public class BukkitWorldInfoProvider implements WorldInfoProvider {
                 entities += world.getEntities().size();
                 Chunk[] chunksArray = world.getLoadedChunks();
                 for (Chunk chunk : chunksArray) {
-                    tileEntities += chunk.getTileEntities().length;
+                    BlockState[] tileEntitiesArray = chunk.getTileEntities();
+                    tileEntities += tileEntitiesArray != null ? tileEntitiesArray.length : 0;
                 }
                 chunks += chunksArray.length;
             }

--- a/spark-bukkit/src/main/java/me/lucko/spark/bukkit/BukkitWorldInfoProvider.java
+++ b/spark-bukkit/src/main/java/me/lucko/spark/bukkit/BukkitWorldInfoProvider.java
@@ -70,12 +70,21 @@ public class BukkitWorldInfoProvider implements WorldInfoProvider {
                 chunks += world.getChunkCount();
             } else {
                 entities += world.getEntities().size();
+
                 Chunk[] chunksArray = world.getLoadedChunks();
+                int nullChunks = 0;
+
                 for (Chunk chunk : chunksArray) {
+                    if (chunk == null) {
+                        ++nullChunks;
+                        continue;
+                    }
+
                     BlockState[] tileEntitiesArray = chunk.getTileEntities();
                     tileEntities += tileEntitiesArray != null ? tileEntitiesArray.length : 0;
                 }
-                chunks += chunksArray.length;
+
+                chunks += chunksArray.length - nullChunks;
             }
         }
 

--- a/spark-common/src/main/java/me/lucko/spark/common/sampler/BackgroundSamplerManager.java
+++ b/spark-common/src/main/java/me/lucko/spark/common/sampler/BackgroundSamplerManager.java
@@ -31,6 +31,7 @@ public class BackgroundSamplerManager {
     private static final String OPTION_ENABLED = "backgroundProfiler";
     private static final String OPTION_ENGINE = "backgroundProfilerEngine";
     private static final String OPTION_INTERVAL = "backgroundProfilerInterval";
+    private static final String OPTION_THREAD_GROUPER = "backgroundProfilerThreadGrouper";
 
     private static final String MARKER_FAILED = "_marker_background_profiler_failed";
 
@@ -101,10 +102,23 @@ public class BackgroundSamplerManager {
     private void startSampler() {
         boolean forceJavaEngine = this.configuration.getString(OPTION_ENGINE, "async").equals("java");
 
+        ThreadGrouper threadGrouper;
+        switch (this.configuration.getString(OPTION_THREAD_GROUPER, "byPool")) {
+            case "asOne":
+                threadGrouper = ThreadGrouper.AS_ONE;
+                break;
+            case "byName":
+                threadGrouper = ThreadGrouper.BY_NAME;
+                break;
+            default:
+                threadGrouper = ThreadGrouper.BY_POOL;
+                break;
+        }
+
         Sampler sampler = new SamplerBuilder()
                 .background(true)
                 .threadDumper(this.platform.getPlugin().getDefaultThreadDumper())
-                .threadGrouper(ThreadGrouper.BY_POOL)
+                .threadGrouper(threadGrouper)
                 .samplingInterval(this.configuration.getInteger(OPTION_INTERVAL, 10))
                 .forceJavaSampler(forceJavaEngine)
                 .start(this.platform);

--- a/spark-common/src/main/java/me/lucko/spark/common/sampler/BackgroundSamplerManager.java
+++ b/spark-common/src/main/java/me/lucko/spark/common/sampler/BackgroundSamplerManager.java
@@ -24,7 +24,10 @@ import me.lucko.spark.common.SparkPlatform;
 import me.lucko.spark.common.platform.PlatformInfo;
 import me.lucko.spark.common.util.Configuration;
 
+import java.util.Arrays;
+import java.util.Set;
 import java.util.logging.Level;
+import java.util.stream.Collectors;
 
 public class BackgroundSamplerManager {
 
@@ -32,6 +35,7 @@ public class BackgroundSamplerManager {
     private static final String OPTION_ENGINE = "backgroundProfilerEngine";
     private static final String OPTION_INTERVAL = "backgroundProfilerInterval";
     private static final String OPTION_THREAD_GROUPER = "backgroundProfilerThreadGrouper";
+    private static final String OPTION_THREAD_DUMPER = "backgroundProfilerThreadDumper";
 
     private static final String MARKER_FAILED = "_marker_background_profiler_failed";
 
@@ -101,9 +105,11 @@ public class BackgroundSamplerManager {
 
     private void startSampler() {
         boolean forceJavaEngine = this.configuration.getString(OPTION_ENGINE, "async").equals("java");
+        String grouperName = this.configuration.getString(OPTION_THREAD_GROUPER, "byPool");
+        String dumperName = this.configuration.getString(OPTION_THREAD_DUMPER, "default");
 
         ThreadGrouper threadGrouper;
-        switch (this.configuration.getString(OPTION_THREAD_GROUPER, "byPool")) {
+        switch (grouperName) {
             case "asOne":
                 threadGrouper = ThreadGrouper.AS_ONE;
                 break;
@@ -115,13 +121,28 @@ public class BackgroundSamplerManager {
                 break;
         }
 
+        ThreadDumper threadDumper;
+        switch (dumperName) {
+            case "default":
+                threadDumper = this.platform.getPlugin().getDefaultThreadDumper();
+                break;
+            case "*":
+            case "all":
+                threadDumper = ThreadDumper.ALL;
+                break;
+            default:
+                Set<String> threadNames = Arrays.stream(dumperName.split(",")).collect(Collectors.toSet());
+                threadDumper = new ThreadDumper.Specific(threadNames);
+                break;
+        }
+
         Sampler sampler = new SamplerBuilder()
-                .background(true)
-                .threadDumper(this.platform.getPlugin().getDefaultThreadDumper())
-                .threadGrouper(threadGrouper)
-                .samplingInterval(this.configuration.getInteger(OPTION_INTERVAL, 10))
-                .forceJavaSampler(forceJavaEngine)
-                .start(this.platform);
+              .background(true)
+              .threadDumper(threadDumper)
+              .threadGrouper(threadGrouper)
+              .samplingInterval(this.configuration.getInteger(OPTION_INTERVAL, 10))
+              .forceJavaSampler(forceJavaEngine)
+              .start(this.platform);
 
         this.platform.getSamplerContainer().setActiveSampler(sampler);
     }

